src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java

```diff
@@ -0,0 +1,976 @@
+                    emitConst(null);
+                    emitStore(BasicType.L, scopeLocal); // need to initialize all scope locals here in case an exception occurs
+                }
+            }
+            scopeSlots = Arrays.copyOf(initialScopeSlots, numScopes); // fit to size
+            curScopeLocalIdx = 0; // used from emitGetInput
+        }
+
+        // create a Binding.Context for this call
+        if (callingSequence.allocationSize() != 0) {
+            emitConst(callingSequence.allocationSize());
+            emitInvokeStatic(Binding.Context.class, "ofBoundedAllocator", OF_BOUNDED_ALLOCATOR_DESC);
+        } else if (callingSequence.forDowncall()) {
+            emitGetStatic(Binding.Context.class, "DUMMY", BINDING_CONTEXT_DESC);
+        } else {
+            emitInvokeStatic(Binding.Context.class, "ofSession", OF_SESSION_DESC);
+        }
+        CONTEXT_IDX = newLocal(BasicType.L);
+        emitStore(BasicType.L, CONTEXT_IDX);
+
+        // in case the call needs a return buffer, allocate it here.
+        // for upcalls the VM wrapper stub allocates the buffer.
+        if (callingSequence.needsReturnBuffer() && callingSequence.forDowncall()) {
+            emitLoadInternalAllocator();
+            emitAllocateCall(callingSequence.returnBufferSize(), 1);
+            RETURN_BUFFER_IDX = newLocal(BasicType.L);
+            emitStore(BasicType.L, RETURN_BUFFER_IDX);
+        }
+
+        Label tryStart = new Label();
+        Label tryEnd = new Label();
+        Label catchStart = new Label();
+
+        mv.visitLabel(tryStart);
+
+        // stack to keep track of types on the bytecode stack between bindings.
+        // this is needed to e.g. emit the right DUP instruction,
+        // but also used for type checking.
+        typeStack = new ArrayDeque<>();
+        // leaf arg types are the types of the args passed to the leaf handle.
+        // these are collected from VM_STORE instructions for downcalls, and
+        // recipe outputs for upcalls (see uses emitSetOutput for both)
+        leafArgTypes = new ArrayList<>();
+        paramIndex = 1; // +1 to skip SegmentAllocator or MethodHandle
+        for (int i = 0; i < callingSequence.argumentBindingsCount(); i++) {
+            if (callingSequence.forDowncall()) {
+                // for downcalls, recipes have an input value, which we set up here
+                if (callingSequence.needsReturnBuffer() && i == 0) {
+                    assert RETURN_BUFFER_IDX != -1;
+                    emitLoad(BasicType.L, RETURN_BUFFER_IDX);
+                    pushType(MemorySegment.class);
+                } else {
+                    emitGetInput();
+                }
+            }
+
+            // emit code according to binding recipe
+            doBindings(callingSequence.argumentBindings(i));
+
+            if (callingSequence.forUpcall()) {
+                // for upcalls, recipes have a result, which we handle here
```

Comment:
I'm not sure if there is anything actionable here?

I've thought in the past that it might be nice to have `GetArgument`/`SetArgument` and `GetReturnValue`/`SetReturnValue` binding operators as well, to make the inputs/outputs more explicit in the recipe. But, it doesn't seem like that would make things _much_ better...

Reply:
It's not quite that simple since a binding recipe for a single parameter can have multiple VMStores for instance if a struct is decomposed into multiple values.

It can be done by pulling the binding loops up to the `specialize` method, and have if statements for VMStore and VMLoad, like this:

    for (Binding binding : callingSequence.argumentBindings(i)) {
        if (binding instanceof Binding.VMStore vms && callingSequence.forDowncall()) {
            emitSetOutput(vms.type());
        } else if (binding instanceof Binding.VMLoad && callingSequence.forUpcall()) {
            emitGetInput();
        } else {
            doBinding(binding);
        }
    }

And for returns:

    for (Binding binding : callingSequence.returnBindings()) {
        if (binding instanceof Binding.VMLoad vml && callingSequence.forDowncall()) {
            if (!callingSequence.needsReturnBuffer()) {
                emitRestoreReturnValue(vml.type());
            } else {
                emitReturnBufferLoad(vml);
            }
        } else if (binding instanceof Binding.VMStore vms && callingSequence.forUpcall()) {
            if (!callingSequence.needsReturnBuffer()) {
                emitSaveReturnValue(vms.type());
            } else {
                emitReturnBufferStore(vms);
            }
        } else {
            doBinding(binding);
        }
    }

But, maybe that's better?

Reply:
I think someone might look at this and think "why aren't these bindings handled by `doBinding`"?